node/src/test/scala/coop/rchain/node/hardening/Hardening.scala

```diff
@@ -0,0 +1,524 @@
+                    }
+                    .tapEval {
+                      case (b, nextLayerNum, balances) =>
+                        verifyBalances(balances, b.body.state.postStateHash) >> {
+                          val usersInvolved =
+                            balances.filterKeys(!validatorUsers.contains(_)).filter {
+                              case (_, (_, txs)) => txs.nonEmpty
+                            }
+                          val paymentsRejected = balances.flatMap(_._2._2).toSet.count(_.rejected)
+                          val paymentsIn       = balances.flatMap(_._2._2).toSet.count(v => !v.rejected)
+
+                          // each tx has charge
+                          assert(
+                            paymentsRejected % 2 == 0,
+                            balances.flatMap(_._2._2).filter(_.rejected)
+                          )
+                          assert(
+                            paymentsIn % 2 == 0,
+                            balances.flatMap(_._2._2).filter(v => !v.rejected)
+                          )
+                          val txRejected = paymentsRejected / 2
+                          val txIn       = paymentsIn / 2
+                          val txTotal    = txIn + txRejected
+
+                          println(
+                            s"""Layer ${nextLayerNum} done. Network stats:
+                               | Validators num: $validatorsNum
+                               | Users total: $usersNum
+                               | Users involved in TX: ${usersInvolved.size}
+                               | Avg payments per block: ${txTotal.toFloat / nextLayerNum / validatorsNum}
+                               | Payments rejected: ${txRejected} (${txRejected.toFloat / txTotal * 100} %)
+                               | Payments accepted: ${txIn} (${txIn.toFloat / txTotal * 100} %)
+                               |""".stripMargin
+                          ).pure
+                        }
+                    }
+                    .map(_.asLeft[(BlockMessage, BalanceSheet)])
+
+                  if (layerNum == mergesNum)
+                    (baseBlock, balanceSheetAcc)
+                      .asRight[(BlockMessage, Int, BalanceSheet)]
+                      .pure[Task]
+                  else newLayer
+              }
+        } yield ()
+      }
+    }
+  }
+}
+
+object Hardening {
+
+  val usersNum                 = 10
+  val validatorsNum            = 5
+  val PER_VALIDATOR_VAULT_ADDR = "validator-pos-vault"
+  val MAX_PAYMENTS_PER_BLOCK   = 5
+  val (minTxAmt, maxTxAmt)     = (10, 100)
+
+  val users          = User.mkRandom(usersNum)
+  val randomPayments = Payment.randomPayments(users, minTxAmt, maxTxAmt)
+  val randomPaymentBatches =
+    Iterator.continually(randomPayments.take(1 + Random.nextInt(MAX_PAYMENTS_PER_BLOCK - 1)).toList)
+
+//    This is how you can make predictable transfers across the network.
```

Comment:
Here most of test variables are defined.
`usersNum` - number of users created in genesis, all of them will be randomly picked as source and target for transfer.
`validatorsNum` number of validators. DAG is formed according to the following logic. Validators `1` - `validatorsNum` create blocks simultaneously, then validator `0` merge them and creates its own block.
`MAX_PAYMENTS_PER_BLOCK` -  each block contain from 1 to this number of random REV transfers 
`randomPaymentBatches ` this is iterator for batches of random transactions put in each block.